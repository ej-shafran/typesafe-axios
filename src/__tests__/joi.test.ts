import { taxios } from "..";

import Joi from "joi";
import nock from "nock";
import * as UTILS from "./__utils__";

const SCHEMA = Joi.object({
  id: Joi.string().required(),
  username: Joi.string().required(),
  email: Joi.string().required(),
  address: Joi.object({
    street: Joi.string().required(),
    zipcode: Joi.string(),
  }),
  followers: Joi.array().items(
    Joi.object({
      id: Joi.string().required(),
      role: Joi.string().valid("admin", "guest"),
    })
  ),
});

describe("joi interop", () => {
  it("passes along data with the correct type", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);
    const { data } = await taxios(SCHEMA).get(UTILS.API);

    expect(data.value).toEqual(UTILS.VALID);
  });

  it("throws a validation error when the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    const { data } = await taxios(SCHEMA).get(UTILS.API);
    expect(JSON.stringify(data.error, null, 2)).toMatchSnapshot();
  });
});
