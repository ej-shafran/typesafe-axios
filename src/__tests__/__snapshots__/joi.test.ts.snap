// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joi interop throws a validation error when the data is invalid 1`] = `
"{
  "_original": {
    "id": "SOME_ID",
    "username": "SOME_USERNAME",
    "address": "ABC123 St.",
    "followers": []
  },
  "details": [
    {
      "message": "\\"email\\" is required",
      "path": [
        "email"
      ],
      "type": "any.required",
      "context": {
        "label": "email",
        "key": "email"
      }
    }
  ]
}"
`;
